---
title: "PCA_UMAP"
author: "N. Alcala"
date: "9/6/2022"
output: 
  html_document: 
    keep_md: yes
---


# R Code to run PCA and UMAP on TCGA expression data

## load libraries 
```{r libraries, message=FALSE}
library(tidyverse)
library(SummarizedExperiment)
library(patchwork)
library(umap)
library(ade4)
```

## Load expression data from TCGA
```{r load, message=FALSE}
load("data/TCGA-MESOTranscriptome_ProfilingFri_Jul_29_11:39:27_2022.RData")
MESO.Rnaseq.SE = data
load("data/TCGA-BRCATranscriptome_ProfilingFri_Jul_29_12:11:29_2022.RData")
BRCA.Rnaseq.SE = data
```

### Get expression in TPM
```{r tpm, message=FALSE}
MESO.Rnaseq.expr.tpm = assay(MESO.Rnaseq.SE,"tpm_unstrand")
BRCA.Rnaseq.expr.tpm = assay(BRCA.Rnaseq.SE,"tpm_unstrand")
```

## Analysis
### PCA of log(tpm+1) expression of 5000 most variable genes
```{r pca, message=FALSE}
pca.MESO = dudi.pca(t(log(MESO.Rnaseq.expr.tpm[order(apply(MESO.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1)),scannf = F,nf=10)
pca.BRCA = dudi.pca(t(log(BRCA.Rnaseq.expr.tpm[order(apply(BRCA.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1)),scannf = F,nf=10)
```

### UMAP of log(tpm+1) expression of 5000 most variable genes
```{r umap, message=FALSE}
umap.MESO = umap(t(log(MESO.Rnaseq.expr.tpm[order(apply(MESO.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1)))
umap.BRCA = umap(t(log(BRCA.Rnaseq.expr.tpm[order(apply(BRCA.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1)))
```

### plot
We create tibble objects for plotting
```{r tibbles, message=FALSE}
embeddings.MESO.tib = tibble(PCA=pca.MESO$li, UMAP = umap.MESO$layout, data=as.data.frame(colData(MESO.Rnaseq.SE) ) )
embeddings.BRCA.tib = tibble(PCA=pca.BRCA$li, UMAP = umap.BRCA$layout, data=as.data.frame(colData(BRCA.Rnaseq.SE) ) )
```

Plot first 2 PCA axes. For MESO, we use the histopathological types as colors; for BRCA, we use the molecular subtypes
```{r pcaplot, message=FALSE}
ggMESO_PCA  = ggplot(embeddings.MESO.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=data$primary_diagnosis)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
ggBRCA_PCA  = ggplot(embeddings.BRCA.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=data$paper_BRCA_Subtype_PAM50)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
```

We plot the UMAP
```{r umapplot, message=FALSE}
ggMESO_UMAP = ggplot(embeddings.MESO.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=data$primary_diagnosis)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
ggBRCA_UMAP = ggplot(embeddings.BRCA.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=data$paper_BRCA_Subtype_PAM50)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
```

We arrange the plots with patchwork
```{r allplots, message=FALSE}
((ggMESO_PCA+guides(color="none"))+ ggMESO_UMAP)/((ggBRCA_PCA+guides(color="none"))+ ggBRCA_UMAP)
```

## Session Info 
```{r info}
sessionInfo()
```