---
title: "PCA_UMAP_AE_benchmark"
author: "N. Alcala"
date: "9/6/2022"
output: 
  html_document: 
    keep_md: yes
---


# R Code to compare embeddings on TCGA expression data

## load libraries 
```{r libraries, message=FALSE}
library(tidyverse)
library(SummarizedExperiment)
library(patchwork)
library(umap)
library(ade4)
library(keras)
library(cluster)
library(readxl)
```

## Load expression data from TCGA
```{r load, message=FALSE}
load("AIML-WG/data/TCGA-MESOTranscriptome_ProfilingFri_Jul_29_11:39:27_2022.RData")
MESO.Rnaseq.SE = data
# clean histology
MESO.Rnaseq.SE$primary_diagnosis[MESO.Rnaseq.SE$primary_diagnosis=="Epithelioid mesothelioma, malignant"] = "MME"
MESO.Rnaseq.SE$primary_diagnosis[MESO.Rnaseq.SE$primary_diagnosis=="Mesothelioma, biphasic, malignant"] = "MMB"
MESO.Rnaseq.SE$primary_diagnosis[MESO.Rnaseq.SE$primary_diagnosis%in%c("Mesothelioma, malignant","Fibrous mesothelioma, malignant")] = NA
table(MESO.Rnaseq.SE$primary_diagnosis)

load("AIML-WG/data/TCGA-BRCATranscriptome_ProfilingFri_Jul_29_12:11:29_2022.RData")
BRCA.Rnaseq.SE = data
```
### load additional clinical data from TCGA papers
```{r }
MESO_S1 = read_xlsx("AIML-WG/data/TCGA_TableS1.xlsx",sheet=2)
# match barcodes 
MESO_S1$barcode = MESO.Rnaseq.SE$barcode[sapply(MESO_S1$TCGA_barcode, function(x) grep(x = MESO.Rnaseq.SE$barcode,pattern = x) )]
MESOorder = sapply(MESO.Rnaseq.SE$barcode,function(x){res=which(MESO_S1$barcode==x);if(length(res)==0){res=NA};return(res)})
# remove samples not reported in study
MESO.Rnaseq.SE = MESO.Rnaseq.SE[,!is.na(MESOorder)]
MESO_S1 = MESO_S1[MESOorder[!is.na(MESOorder)],]
```

### make survival groups
```{r }
## MESO
MESO.Rnaseq.SE$survival_group = cut(MESO.Rnaseq.SE$days_to_death,breaks = quantile(MESO.Rnaseq.SE$days_to_death,c(0,0.33,0.67,1),na.rm=T))
MESO.Rnaseq.SE$survival_group[which(as.numeric(MESO.Rnaseq.SE$survival_group)<3 & MESO.Rnaseq.SE$vital_status=="Alive")] = NA

## BRCA
BRCA.Rnaseq.SE$paper_days_to_death = as.numeric(BRCA.Rnaseq.SE$paper_days_to_death)
BRCA.Rnaseq.SE$survival_group = cut(BRCA.Rnaseq.SE$paper_days_to_death,breaks = quantile(BRCA.Rnaseq.SE$paper_days_to_death,c(0,0.33,0.67,1),na.rm=T))
BRCA.Rnaseq.SE$survival_group[which(as.numeric(BRCA.Rnaseq.SE$survival_group)<3 & BRCA.Rnaseq.SE$paper_vital_status=="Alive")] = NA
```


### Get expression in TPM
```{r tpm, message=FALSE}
MESO.Rnaseq.expr.tpm = assay(MESO.Rnaseq.SE,"tpm_unstrand")
BRCA.Rnaseq.expr.tpm = assay(BRCA.Rnaseq.SE,"tpm_unstrand")

MESO.Rnaseq.expr.tpm.logmat = t(log(MESO.Rnaseq.expr.tpm[order(apply(MESO.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1))
BRCA.Rnaseq.expr.tpm.logmat = t(log(BRCA.Rnaseq.expr.tpm[order(apply(BRCA.Rnaseq.expr.tpm,1,var),decreasing = T)[1:5000],]+1))
```

## Analysis
Note: use bootstrapping to get CIs.

### PCA of log(tpm+1) expression of 5000 most variable genes
```{r pca, message=FALSE}
pca.MESO = dudi.pca(MESO.Rnaseq.expr.tpm.logmat,scannf = F,nf=10)
pca.BRCA = dudi.pca(BRCA.Rnaseq.expr.tpm.logmat,scannf = F,nf=10)
```

### UMAP of log(tpm+1) expression of 5000 most variable genes
```{r umap, message=FALSE}
umap.MESO = umap(MESO.Rnaseq.expr.tpm.logmat)
umap.BRCA = umap(BRCA.Rnaseq.expr.tpm.logmat)
```
Note: play with parameters.

### Autoencoder of of log(tpm+1) expression of 5000 most variable genes
```{r AE, message=FALSE}
# set models
modelMESO <- keras_model_sequential()
modelBRCA <- keras_model_sequential()
modelMESO %>%
  layer_dense(units = 50, activation = "linear", input_shape = 5000) %>%
  layer_dense(units = 2, activation = "linear", name = "bottleneck") %>%
  layer_dense(units = 50, activation = "linear") %>%
  layer_dense(units = 5000)
modelBRCA %>%
  layer_dense(units = 50, activation = "linear", input_shape = 5000) %>%
  layer_dense(units = 2, activation = "linear", name = "bottleneck") %>%
  layer_dense(units = 50, activation = "linear") %>%
  layer_dense(units = 5000)
# view model layers
summary(modelMESO)

# compile model
modelMESO %>% compile(loss = "mean_squared_error", optimizer = "rmsprop")
modelBRCA %>% compile(loss = "mean_squared_error", optimizer = "rmsprop")

# fit model
modelMESO %>% fit(x = MESO.Rnaseq.expr.tpm.logmat, y = MESO.Rnaseq.expr.tpm.logmat, epochs = 200,verbose = 1)
modelBRCA %>% fit(x = BRCA.Rnaseq.expr.tpm.logmat, y = BRCA.Rnaseq.expr.tpm.logmat, epochs = 200,verbose = 1)

# evaluate the performance of the model
mse.aeMESO <- evaluate(modelMESO, MESO.Rnaseq.expr.tpm.logmat, MESO.Rnaseq.expr.tpm.logmat)
mse.aeBRCA <- evaluate(modelBRCA, BRCA.Rnaseq.expr.tpm.logmat, BRCA.Rnaseq.expr.tpm.logmat)
mse.aeMESO
mse.aeBRCA

intermediate_layer_modelMESO <- keras_model(inputs = modelMESO$input, outputs = get_layer(modelMESO, "bottleneck")$output)
intermediate_layer_modelBRCA <- keras_model(inputs = modelBRCA$input, outputs = get_layer(modelBRCA, "bottleneck")$output)
AE.MESO <- predict(intermediate_layer_modelMESO, MESO.Rnaseq.expr.tpm.logmat)
AE.BRCA <- predict(intermediate_layer_modelBRCA, BRCA.Rnaseq.expr.tpm.logmat)
```

### plot
We create tibble objects for plotting
```{r tibbles, message=FALSE}
embeddings.MESO.tib = tibble(PCA=pca.MESO$li, UMAP = umap.MESO$layout, AE= AE.MESO, data=as.data.frame(colData(MESO.Rnaseq.SE) ) )
embeddings.BRCA.tib = tibble(PCA=pca.BRCA$li, UMAP = umap.BRCA$layout, AE=AE.BRCA, data=as.data.frame(colData(BRCA.Rnaseq.SE) ) )
```

#### Histopathology
Plot first 2 PCA axes. For MESO, we use the histopathological types as colors; for BRCA, we use the molecular subtypes
```{r pcaplot, message=FALSE}
ggMESO_PCA  = ggplot(embeddings.MESO.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=MESO.Rnaseq.SE$primary_diagnosis)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
ggBRCA_PCA  = ggplot(embeddings.BRCA.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=BRCA.Rnaseq.SE$paper_BRCA_Pathology)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
```

We plot the UMAP
```{r umapplot, message=FALSE}
ggMESO_UMAP = ggplot(embeddings.MESO.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=MESO.Rnaseq.SE$primary_diagnosis)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
ggBRCA_UMAP = ggplot(embeddings.BRCA.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=BRCA.Rnaseq.SE$paper_BRCA_Pathology)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
```

We plot the autoencoders
```{r aeplot, message=FALSE}
ggMESO_AE = ggplot(embeddings.MESO.tib, aes(x=AE[,1],y=AE[,2] ,col=MESO.Rnaseq.SE$primary_diagnosis)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
ggBRCA_AE = ggplot(embeddings.BRCA.tib, aes(x=AE[,1],y=AE[,2] ,col=BRCA.Rnaseq.SE$paper_BRCA_Pathology)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
```

We arrange the plots with patchwork
```{r allplots, message=FALSE}
((ggMESO_PCA+guides(color="none"))+ ggMESO_UMAP+guides(color="none")+ (ggMESO_AE + guides(col=guide_legend(title="Histology"))) )/((ggBRCA_PCA+guides(color="none"))+ ggBRCA_UMAP+guides(color="none")+ (ggBRCA_AE + guides(col=guide_legend(title="Histology"))) )
```

#### Molecular groups
```{r pcaplotM, message=FALSE}
ggMESO_PCA  = ggplot(embeddings.MESO.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=MESO_S1$iCluster_k4.5types)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
ggBRCA_PCA  = ggplot(embeddings.BRCA.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
```

We plot the UMAP
```{r umapplotM, message=FALSE}
ggMESO_UMAP = ggplot(embeddings.MESO.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=MESO_S1$iCluster_k4.5types)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
ggBRCA_UMAP = ggplot(embeddings.BRCA.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
```

We plot the autoencoders
```{r aeplotM, message=FALSE}
ggMESO_AE = ggplot(embeddings.MESO.tib, aes(x=AE[,1],y=AE[,2] ,col=MESO_S1$iCluster_k4.5types)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
ggBRCA_AE = ggplot(embeddings.BRCA.tib, aes(x=AE[,1],y=AE[,2] ,col=BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
```

We arrange the plots with patchwork
```{r allplotsM, message=FALSE}
((ggMESO_PCA+guides(color="none"))+ ggMESO_UMAP+guides(color="none")+ (ggMESO_AE + guides(col=guide_legend(title="Molecular group"))) )/((ggBRCA_PCA+guides(color="none"))+ ggBRCA_UMAP+guides(color="none")+ (ggBRCA_AE + guides(col=guide_legend(title="Molecular group"))) )
```

#### Clinical groups
```{r pcaplotC, message=FALSE}
ggMESO_PCA  = ggplot(embeddings.MESO.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=MESO.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
ggBRCA_PCA  = ggplot(embeddings.BRCA.tib, aes(x=PCA$Axis1,y=PCA$Axis2 ,col=BRCA.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("PC1") + ylab("PC2")
```

We plot the UMAP
```{r umapplotC, message=FALSE}
ggMESO_UMAP = ggplot(embeddings.MESO.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=MESO.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
ggBRCA_UMAP = ggplot(embeddings.BRCA.tib, aes(x=UMAP[,1],y=UMAP[,2] ,col=BRCA.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("UMAP1") + ylab("UMAP2")
```

We plot the autoencoders
```{r aeplotC, message=FALSE}
ggMESO_AE = ggplot(embeddings.MESO.tib, aes(x=AE[,1],y=AE[,2] ,col=MESO.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
ggBRCA_AE = ggplot(embeddings.BRCA.tib, aes(x=AE[,1],y=AE[,2] ,col=BRCA.Rnaseq.SE$survival_group)) + geom_point() + theme_bw() + xlab("AE1") + ylab("AE2")
```

We arrange the plots with patchwork
```{r allplotsC, message=FALSE}
((ggMESO_PCA+guides(color="none"))+ ggMESO_UMAP+guides(color="none")+ (ggMESO_AE + guides(col=guide_legend(title="Survival group"))) )/((ggBRCA_PCA+guides(color="none"))+ ggBRCA_UMAP+guides(color="none")+ (ggBRCA_AE + guides(col=guide_legend(title="Survival group"))) )
```

## Compute metrics
### For histopathological types
```{r sil, message=FALSE}
## MESO
silMESO.PCA  = silhouette(as.numeric(as.factor(MESO.Rnaseq.SE$primary_diagnosis))[!is.na(MESO.Rnaseq.SE$primary_diagnosis)],
           dist = dist(embeddings.MESO.tib$PCA[!is.na(MESO.Rnaseq.SE$primary_diagnosis),]))
silMESO.UMAP = silhouette(as.numeric(as.factor(MESO.Rnaseq.SE$primary_diagnosis))[!is.na(MESO.Rnaseq.SE$primary_diagnosis)],
           dist = dist(embeddings.MESO.tib$UMAP[!is.na(MESO.Rnaseq.SE$primary_diagnosis),]))
silMESO.AE   = silhouette(as.numeric(as.factor(MESO.Rnaseq.SE$primary_diagnosis))[!is.na(MESO.Rnaseq.SE$primary_diagnosis)],
           dist = dist(embeddings.MESO.tib$AE[!is.na(MESO.Rnaseq.SE$primary_diagnosis),]))
## BRCA
silBRCA.PCA  = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Pathology))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology)],
           dist = dist(embeddings.BRCA.tib$PCA[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology),]))
silBRCA.UMAP = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Pathology))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology)],
           dist = dist(embeddings.BRCA.tib$UMAP[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology),]))
silBRCA.AE   = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Pathology))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology)],
           dist = dist(embeddings.BRCA.tib$AE[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Pathology),]))
```

### For molecular groups
```{r silmol, message=FALSE}
## MESO
silMESO.PCA.mol  = silhouette(as.numeric(as.factor(MESO_S1$iCluster_k4.5types)), dist = dist(embeddings.MESO.tib$PCA))
silMESO.UMAP.mol = silhouette(as.numeric(as.factor(MESO_S1$iCluster_k4.5types)), dist = dist(embeddings.MESO.tib$UMAP))
silMESO.AE.mol   = silhouette(as.numeric(as.factor(MESO_S1$iCluster_k4.5types)), dist = dist(embeddings.MESO.tib$AE))
## BRCA
silBRCA.PCA.mol  = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)],
           dist = dist(embeddings.BRCA.tib$PCA[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50),]))
silBRCA.UMAP.mol = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)],
           dist = dist(embeddings.BRCA.tib$UMAP[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50),]))
silBRCA.AE.mol   = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50))[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50)],
           dist = dist(embeddings.BRCA.tib$AE[!is.na(BRCA.Rnaseq.SE$paper_BRCA_Subtype_PAM50),]))
```

### For clinical variables
```{r silclin, message=FALSE}
## MESO
silMESO.PCA.clin  = silhouette(as.numeric(MESO.Rnaseq.SE$survival_group)[!is.na(MESO.Rnaseq.SE$survival_group)], dist = dist(embeddings.MESO.tib$PCA[!is.na(MESO.Rnaseq.SE$survival_group),]))
silMESO.UMAP.clin = silhouette(as.numeric(MESO.Rnaseq.SE$survival_group)[!is.na(MESO.Rnaseq.SE$survival_group)], dist = dist(embeddings.MESO.tib$UMAP[!is.na(MESO.Rnaseq.SE$survival_group),]))
silMESO.AE.clin   = silhouette(as.numeric(MESO.Rnaseq.SE$survival_group)[!is.na(MESO.Rnaseq.SE$survival_group)], dist = dist(embeddings.MESO.tib$AE[!is.na(MESO.Rnaseq.SE$survival_group),]))

## BRCA
silBRCA.PCA.clin  = silhouette(as.numeric(BRCA.Rnaseq.SE$survival_group)[!is.na(BRCA.Rnaseq.SE$survival_group)],
           dist = dist(embeddings.BRCA.tib$PCA[!is.na(BRCA.Rnaseq.SE$survival_group),]))
silBRCA.UMAP.clin = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$survival_group))[!is.na(BRCA.Rnaseq.SE$survival_group)],
           dist = dist(embeddings.BRCA.tib$UMAP[!is.na(BRCA.Rnaseq.SE$survival_group),]))
silBRCA.AE.clin   = silhouette(as.numeric(as.factor(BRCA.Rnaseq.SE$survival_group))[!is.na(BRCA.Rnaseq.SE$survival_group)],
           dist = dist(embeddings.BRCA.tib$AE[!is.na(BRCA.Rnaseq.SE$survival_group),]))
```


### Plot metrics
```{r silplot, message=FALSE}
sill = tibble(embedding=rep(c("PCA","UMAP","AE"),2*3),variable=rep(c("Histological types","Molecular groups","Survival groups"),each=3*2), 
              Cohort = rep(rep(c("MESO","BRCA"),each=3),3),
              silhouette=c(mean(silMESO.PCA[,3]),mean(silMESO.UMAP[,3]),mean(silMESO.AE[,3]),
                           mean(silBRCA.PCA[,3]),mean(silBRCA.UMAP[,3]),mean(silBRCA.AE[,3]),
                           mean(silMESO.PCA.mol[,3]),mean(silMESO.UMAP.mol[,3]),mean(silMESO.AE.mol[,3]),
                           mean(silBRCA.PCA.mol[,3]),mean(silBRCA.UMAP.mol[,3]),mean(silBRCA.AE.mol[,3]),
                           mean(silMESO.PCA.clin[,3]),mean(silMESO.UMAP.clin[,3]),mean(silMESO.AE.clin[,3]),
                           mean(silBRCA.PCA.clin[,3]),mean(silBRCA.UMAP.clin[,3]),mean(silBRCA.AE.clin[,3])
                           ))

ggplot(sill, aes(x=embedding,y=silhouette,fill=Cohort) ) + geom_bar(stat="identity",beside=T,position=position_dodge()) + 
  theme_bw() +ylab("Mean silhouette width") + facet_grid(.~variable)
```

Note: implement other metrics and find a way to summarize it.

## Session Info 
```{r info}
sessionInfo()
```